### ISSUE TITLE
Connection Drops: Unstable inbound listener config when multiple TCP services map to the same Pod and port

### ISSUE THREAD

**Issue by jtakamine** [istio/istio#18684]
**Bug description**
We have a use case where there are two TCP services (say, ServiceA and ServiceB) that map to the same Pod and port. This is [recognized as a conflict](https://github.com/istio/istio/blob/1.3.4/pilot/pkg/networking/core/v1alpha3/listener.go#L529-L538) when building the inbound listener map for the Pod (in itself, this isn't an issue since both TCP services expect the same protocol). The problem is that when this inbound listener map is constructed, the service that "wins" the conflict is flapping between ServiceA and ServiceB. So when the inbound listener map is pushed down to the Pod sidecar, listeners get destroyed and recreated, which in turn destroys in-flight connections. Since this service is a database, and we use connection pooling, this is disruptive.

**Expected behavior**
At a high-level, the connections should not be destroyed when the listener map is pushed down to the Pod sidecar, since nothing _material_ has changed.

I can see that being achieved in a few different ways, but I'm not sure what the best approach would be. It seems like the simplest mitigation here would be to sort the [list of node ServiceInstances](https://github.com/istio/istio/blob/1.3.4/pilot/pkg/networking/core/v1alpha3/listener.go#L345) prior to generating the listener map so that the result is stable even in the face of conflicts.

**Steps to reproduce the bug**

1. Create 2 TCP services (prefix port names with "tcp-") that have selectors that match the same Pod and port
2. Periodically dump the listener config for the Pod (over a period of ~30min-1hr)
3. Observe that the listener config changes non-deterministically. Sometimes it contains a listener for one of the services and sometimes the other. This changing listener config causes connections to get dropped.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
```
$ istioctl version --remote
client version: unknown
citadel version: 1.3.4
galley version: 1.3.4
galley version: 1.3.4
pilot version: unknown
pilot version: unknown
pilot version: unknown
policy version: 1.3.4
policy version: 1.3.4
policy version: 1.3.4
sidecar-injector version: 1.3.4
sidecar-injector version: 1.3.4
telemetry version: 1.3.4
telemetry version: 1.3.4
telemetry version: 1.3.4
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.7", GitCommit:"8fca2ec50a6133511b771a11559e24191b1aa2b4", GitTreeState:"clean", BuildDate:"2019-09-18T14:47:22Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-gke.5", GitCommit:"5976cae75ee00b052334bfbbd28c4df4f40f1489", GitTreeState:"clean", BuildDate:"2019-10-05T00:04:27Z", GoVersion:"go1.11.13b4", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by howardjohn** [comment]
@nrjpoddar @yangminzhu

**Comment by nicktrav** [comment]
> It seems like the simplest mitigation here would be to sort the list of node ServiceInstances prior to generating the listener map so that the result is stable even in the face of conflicts.

In terms of the sorting, it seems like a similar approach was taken in #18091, which was also working around instability in the inbound listeners.

There's also #18089, which seems relevant here in terms of the listener change triggering reload, which is severing connections.

**Comment by hzxuzhonghu** [comment]
We don't have test cases cover this scenario: pods belong to multi services, the behavior is unknown for us.  IIRC, previously we announced this.

**Comment by nicktrav** [comment]
> IIRC, previously we announced this.

@hzxuzhonghu - Would be interested to see where this is documented. The documentation explicitly calls out the fact that a Pod can be belong to multiple services with the same port, as long as the protocol is the same:

> Service association: A pod must belong to at least one Kubernetes service even if the pod does NOT expose any port. If a pod belongs to multiple Kubernetes services, the services cannot use the same port number for different protocols, for instance HTTP and TCP.

https://istio.io/docs/setup/additional-setup/requirements/

**Comment by hzxuzhonghu** [comment]
This is the first time i see this doc. But try to think of a case for headless tcp service, we now build a listener per instance. In this case there must be duplicate outbound listeners.

**Comment by nicktrav** [comment]
Just to make sure we're on the same page here:

> try to think of a case for headless tcp service

The setup in this case is _not_ using headless services.

> In this case there must be duplicate outbound listeners.

Note that we're only having an issue with _inbound_ listeners.

**Comment by howardjohn** [comment]
I think this is a valid issue and sorting the ServiceInstances seems like a reasonable approach. We should have some deterministic way of selecting things when there are conflicts. In other cases, like DRs, VSs etc we go based on creation date -- that probably doesn't work well here though

**Comment by nrjpoddar** [comment]
I agree, it feels like we are jumping to conclusion a bit early. @nicktrav can you share the following config/data from your cluster:

1) Services that cause this behavior
2) Before and after listener config dump of the affected sidecar.

Thanks!

**Comment by jtakamine** [comment]
Thanks for looking into this! I've attached the requested info below:
[cluster-info.txt](https://github.com/istio/istio/files/3817970/cluster-info.txt)

**Comment by jtakamine** [comment]
As a side note, while I think sorting the ServiceInstances would mitigate this issue for us, I'm wondering if there are any other options for implementing our use case without causing inbound listener conflicts.

It _seems_ like these inbound listener conflicts don't in themselves cause problems (if they're deterministically resolved). But I'm slightly worried that either now or in the future, throwing away one of our services when creating the inbound listener map will be problematic. Would appreciate your thoughts on this.

Thanks!

**Comment by pbohman** [comment]
I also posted this problem in: https://discuss.istio.io/t/pilot-produces-non-deterministic-envoy-listeners-when-service-endpoints-overlap/4157

> I agree, it feels like we are jumping to conclusion a bit early. nicktrav can you share the following config/data from your cluster:
Services that cause this behavior

https://gist.github.com/pbohman/2c1d5a488418c33b847c6f548e60b8c0#file-tcp-echo-yaml

>Before and after listener config dump of the affected sidecar. I'm assuming the diff will be in the inbound section and not VirtualInbound?

https://gist.github.com/pbohman/da3e4158bb89110aae7413a4e37c36a5/revisions#diff-f435f4bdade6947212aa0ace7f3566ba

Can't speak to the announcement or user docs, but found this comment in the code when I sorted the proxy service instances.
 
> ... multiple services may be implemented by the same physical port number,
though with a different ServicePort and NetworkEndpoint for each.  If any of these overlapping
services are not HTTP or H2-based, behavior is undefined, since the listener may not be able to determine the intended destination of a connection without a Host header on the request.
 
https://github.com/istio/istio/blob/master/pilot/pkg/model/service.go#L454
 
FWIW, sorting the listeners fixed the problem.
 
```diff
diff --git a/pilot/pkg/model/context.go b/pilot/pkg/model/context.go
index a1fc163..52b117a 100644
--- a/pilot/pkg/model/context.go
+++ b/pilot/pkg/model/context.go
@@ -20,6 +20,7 @@ import (
        "fmt"
        "net"
        "regexp"
+       "sort"
        "strconv"
        "strings"
@@ -473,6 +474,10 @@ func (node *Proxy) SetServiceInstances(env *Environment) error {
                return err
        }
+       sort.SliceStable(instances, func(i, j int) bool {
+               return instances[i].Service.Hostname < instances[j].Service.Hostname
+       })
+
        node.ServiceInstances = instances
        return nil
}
```

**Comment by nrjpoddar** [comment]
Thanks for all the info @jtakamine & @pbohman. I think there are two or three separate issues that we should think about:

* For multiple services targeting the same workloads, how to generate stable Envoy configuration. The minimum requirement for me is if the operator doesn't change any K8s/istio configuration every workload's Envoy configuration is guaranteed to be stable.
* It looks like the instability is due to metrics metadata i.e. stats prefix for Envoy or mixer listeners which is interesting as in this case it's very difficult to infer operator's intent. As you're using TCP protocol, the inbound has to be configured with `<POD_IP>_<PORT>` and as it is TCP, there is no disambiguation that can be done based on the host name because we are dealing with IP addresses at this level. For any telemetry that we generate, we can only chose one of the many services that are all targeting the same workload. Note that if this was HTTP protocol we wouldn't have any issues as we still have access to the host header.

I would like to understand why this configuration is required in your environments and if it is required what service is preferable for metric attribution. If there's a common pattern, then we can chose one service name over the other.

**Comment by jtakamine** [comment]
In our case, we are running a 3 node CockroachDB cluster. There are two classes of routing that happen here:
1. Routing from one CockroachDB node to another
2. Routing from application SQL clients to the CockroachDB cluster

Say we have 3 nodes (pods) in the cluster: Node1, Node2, Node3. We have the following services defined:

Service1: selects on Node1 only
Service2: selects on Node2 only
Service3: selects on Node3 only
ServiceCluster: selects on Node1, Node2, and Node3

During cluster initialization and other gossiping communication, nodes must be able to individually address each other using Service1, Service2 and Service3. On the other hand, when application clients make SQL connections to the cluster, we want to be able to leverage Istio's more advanced load balancing (including features like locality-awareness), so we have those clients address the "opaque" ServiceCluster.

Note: It may seem like using Istio for node-to-node communication (Service1, 2, 3) isn't really necessary here, but it actually does help us since each CockroachDB node is in a different region and different GKE cluster--so basically, in the inter-node routing case, Istio is helping us do federated service discovery across multiple Kubernetes clusters.

Obviously in this setup, we have multiple services attached to the same pod. In our case, there is a service that is strictly more granular than the other "conflicting" service, so I suppose using that granular service (e.g., Service1) for telemetry makes sense so that we're not losing fidelity.

Hope that's helpful. Let me know if I can clarify or expand on anything. Thanks again for taking a look into this!

**Comment by nrjpoddar** [comment]
This is very helpful. In this case the most specific makes sense, I'm not sure if that principle applies in other scenarios but I think picking the most specific one might be just better than any other algorithm we come up with for sorting or resolving conflict. @howardjohn WDYT?

**Comment by pbohman** [comment]
Our use case is for redis. We have a service per pod for pod-to-pod communication and a primary service that points to the primary redis instance. There is an overlap between the primary and per pod services on the redis port.

You mentioned a difference in stats, but I think there would also be a difference in authorization policies if there were different authz policies for the two services. For us, the biggest impact is services restarting without admin activity.

>There is no disambiguation that can be done based on the host name because we are dealing with IP addresses at this level.

We could consider TLS SNI, but it would only apply for TLS use cases. This works on ingress with TLS passthrough.

**Comment by howardjohn** [comment]
Coming up with the "most specific" one seems challenging. You could have three pods, A/B/C, where A+B are part of svc1 and B+C are part of svc2, which is now pretty hard to make the distinction. For the above case shouldn't a headless service be used?

Note that this applies too all the plugins, not just mixer as @pbohman noted

**Comment by nrjpoddar** [comment]
@pbohman I understand the Authz case, the new Authz APIs are ABAC and as such the merge for them is additive and if any of them meet the criterion the request is allowed. Telemetry is harder to achieve as there’s no merging that’s possible.

For me even though the problem applies to all plugins, some lend more naturally to merge semantics as compared to others where only one can win.

**Comment by nrjpoddar** [comment]
> Coming up with the "most specific" one seems challenging. You could have three pods, A/B/C, where A+B are part of svc1 and B+C are part of svc2, which is now pretty hard to make the distinction. For the above case shouldn't a headless service be used?

@howardjohn Yes, but I was thinking like any conflict resolution approach, we would try to pick the best based on our preferred algorithm and then fallback to a deterministic algorithm guaranteed to give stable results.

Let's do a short term fix of sorting for now so that we don't drop connections. Can one of you file a separate issue/feature request to allow multiple services targeting the same workload port. This is a much bigger change both in code and user experience so we need to have a broader agreement for it.

**Comment by nicktrav** [comment]
> Let's do a short term fix of sorting for now so that we don't drop connections.

Sounds reasonable from our end. We were thinking of forking to do this, but would much prefer this be upstreamed.

One issue we (at least Justin and I) have with this approach is that we're essentially solving the issue of recreating the listener by relying on the implementation detail that the list of services used to created the listeners is sorted in stable way.

While this is fine for now, I think it needs a lot more thought on how to solve more generally / robustly, as relying on an implementation detail is brittle.

> This is a much bigger change both in code and user experience so we need to have a broader agreement for it.

Absolutely agree with you here.

> Can one of you file a separate issue/feature request to allow multiple services targeting the same workload port.

Sure thing. We'll write something up based on what we've observed and what would suit our particular use case.

Feel free to assign to me.

**Comment by wu0407** [comment]
same issue with same tcp port on serviceentry and the only one listener is configed.
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: 192.168.1.33
spec:
  hosts:
  - source-native.xxx.com
  ports:
  - number: 80
    name: tcp
    protocol: TCP
    resolution: DNS
  location: MESH_EXTERNAL
  exportTo:
    - .

---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: 192.168.2.40
spec:
  hosts:
  - test-k8s-prometheus.xxx.com
  ports:
  - number: 80
    name: tcp
    protocol: TCP
    resolution: DNS
  location: MESH_EXTERNAL
  exportTo:
    - .
```
istioctl proxy-config listeners  test-network-nginx-canary-7b46c94cb4-hkgcv --port 80 -o json
```
 { 
      "name":"0.0.0.0_80",
      "address":{ 
         "socketAddress":{ 
            "address":"0.0.0.0",
            "portValue":80
         }
      },
      "filterChains":[ 
         { 
            "filterChainMatch":{ 
               "prefixRanges":[ 
                  { 
                     "addressPrefix":"10.2.0.229",
                     "prefixLen":32
                  }
               ]
            },
            "filters":[ 
               { 
                  "name":"mixer",
                  "typedConfig":{ 
                     "@type":"type.googleapis.com/istio.mixer.v1.config.client.TcpClientConfig",
                     "transport":{ 
                        "networkFailPolicy":{ 
                           "policy":"FAIL_CLOSE",
                           "baseRetryWait":"0.080s",
                           "maxRetryWait":"1s"
                        },
                        "checkCluster":"outbound|9091||istio-policy.istio-system.svc.cluster.local",
                        "reportCluster":"outbound|9091||istio-telemetry.istio-system.svc.cluster.local",
                        "reportBatchMaxEntries":100,
                        "reportBatchMaxTime":"1s"
                     },
                     "mixerAttributes":{ 
                        "attributes":{ 
                           "context.proxy_version":{ 
                              "stringValue":"1.4.0"
                           },
                           "context.reporter.kind":{ 
                              "stringValue":"outbound"
                           },
                           "context.reporter.uid":{ 
                              "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                           },
                           "destination.service.host":{ 
                              "stringValue":"test-k8s-prometheus.xxx.com"
                           },
                           "destination.service.name":{ 
                              "stringValue":"test-k8s-prometheus.xxx.com"
                           },
                           "destination.service.namespace":{ 
                              "stringValue":"default"
                           },
                           "source.namespace":{ 
                              "stringValue":"default"
                           },
                           "source.uid":{ 
                              "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                           }
                        }
                     },
                     "disableCheckCalls":true
                  }
               },
               { 
                  "name":"envoy.tcp_proxy",
                  "typedConfig":{ 
                     "@type":"type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                     "statPrefix":"BlackHoleCluster",
                     "cluster":"BlackHoleCluster"
                  }
               }
            ]
         },
         { 
            "filters":[ 
               { 
                  "name":"mixer",
                  "typedConfig":{ 
                     "@type":"type.googleapis.com/istio.mixer.v1.config.client.TcpClientConfig",
                     "transport":{ 
                        "networkFailPolicy":{ 
                           "policy":"FAIL_CLOSE",
                           "baseRetryWait":"0.080s",
                           "maxRetryWait":"1s"
                        },
                        "checkCluster":"outbound|9091||istio-policy.istio-system.svc.cluster.local",
                        "reportCluster":"outbound|9091||istio-telemetry.istio-system.svc.cluster.local",
                        "reportBatchMaxEntries":100,
                        "reportBatchMaxTime":"1s"
                     },
                     "mixerAttributes":{ 
                        "attributes":{ 
                           "context.proxy_version":{ 
                              "stringValue":"1.4.0"
                           },
                           "context.reporter.kind":{ 
                              "stringValue":"outbound"
                           },
                           "context.reporter.uid":{ 
                              "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                           },
                           "destination.service.host":{ 
                              "stringValue":"test-k8s-prometheus.xxx.com"
                           },
                           "destination.service.name":{ 
                              "stringValue":"test-k8s-prometheus.xxx.com"
                           },
                           "destination.service.namespace":{ 
                              "stringValue":"default"
                           },
                           "source.namespace":{ 
                              "stringValue":"default"
                           },
                           "source.uid":{ 
                              "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                           }
                        }
                     },
                     "disableCheckCalls":true
                  }
               },
               { 
                  "name":"envoy.tcp_proxy",
                  "typedConfig":{ 
                     "@type":"type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
                     "statPrefix":"outbound|80||test-k8s-prometheus.xxx.com",
                     "cluster":"outbound|80||test-k8s-prometheus.xxx.com"
                  }
               }
            ]
         },
         { 
            "filterChainMatch":{ 
               "applicationProtocols":[ 
                  "http/1.0",
                  "http/1.1",
                  "h2"
               ]
            },
            "filters":[ 
               { 
                  "name":"envoy.http_connection_manager",
                  "typedConfig":{ 
                     "@type":"type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
                     "statPrefix":"outbound_0.0.0.0_80",
                     "rds":{ 
                        "configSource":{ 
                           "ads":{ 

                           }
                        },
                        "routeConfigName":"80"
                     },
                     "httpFilters":[ 
                        { 
                           "name":"mixer",
                           "typedConfig":{ 
                              "@type":"type.googleapis.com/istio.mixer.v1.config.client.HttpClientConfig",
                              "transport":{ 
                                 "networkFailPolicy":{ 
                                    "policy":"FAIL_CLOSE",
                                    "baseRetryWait":"0.080s",
                                    "maxRetryWait":"1s"
                                 },
                                 "checkCluster":"outbound|9091||istio-policy.istio-system.svc.cluster.local",
                                 "reportCluster":"outbound|9091||istio-telemetry.istio-system.svc.cluster.local",
                                 "reportBatchMaxEntries":100,
                                 "reportBatchMaxTime":"1s"
                              },
                              "serviceConfigs":{ 
                                 "default":{ 
                                    "disableCheckCalls":true
                                 }
                              },
                              "defaultDestinationService":"default",
                              "mixerAttributes":{ 
                                 "attributes":{ 
                                    "context.proxy_version":{ 
                                       "stringValue":"1.4.0"
                                    },
                                    "context.reporter.kind":{ 
                                       "stringValue":"outbound"
                                    },
                                    "context.reporter.uid":{ 
                                       "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                                    },
                                    "source.namespace":{ 
                                       "stringValue":"default"
                                    },
                                    "source.uid":{ 
                                       "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                                    }
                                 }
                              },
                              "forwardAttributes":{ 
                                 "attributes":{ 
                                    "source.uid":{ 
                                       "stringValue":"kubernetes://test-network-nginx-canary-7b46c94cb4-hkgcv.default"
                                    }
                                 }
                              }
                           }
                        },
                        { 
                           "name":"envoy.cors"
                        },
                        { 
                           "name":"envoy.fault"
                        },
                        { 
                           "name":"envoy.router"
                        }
                     ],
                     "tracing":{ 
                        "operationName":"EGRESS",
                        "clientSampling":{ 
                           "value":100
                        },
                        "randomSampling":{ 
                           "value":1
                        },
                        "overallSampling":{ 
                           "value":100
                        }
                     },
                     "httpProtocolOptions":{ 
                        "acceptHttp10":true
                     },
                     "streamIdleTimeout":"0s",
                     "useRemoteAddress":false,
                     "generateRequestId":true,
                     "upgradeConfigs":[ 
                        { 
                           "upgradeType":"websocket"
                        }
                     ],
                     "normalizePath":true
                  }
               }
            ]
         }
      ],
      "deprecatedV1":{ 
         "bindToPort":false
      },
      "listenerFilters":[ 
         { 
            "name":"envoy.listener.tls_inspector"
         },
         { 
            "name":"envoy.listener.http_inspector"
         }
      ],
      "listenerFiltersTimeout":"0.100s",
      "continueOnListenerFiltersTimeout":true,
      "trafficDirection":"OUTBOUND"
   },
```
```
root@test-network-nginx-canary-7b46c94cb4-hkgcv:/# curl -sv test-k8s-prometheus.corp.cootek.com
* Rebuilt URL to: test-k8s-prometheus.xxx.com/
*   Trying 192.168.2.40...
* TCP_NODELAY set
* Connected to test-k8s-prometheus.xxx.com (192.168.2.40) port 80 (#0)
> GET / HTTP/1.1
> Host: test-k8s-prometheus.xxx.com
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 502 Bad Gateway
< location: http://test-k8s-prometheus.xxx.com/
< date: Wed, 04 Dec 2019 08:17:23 GMT
< server: envoy
< content-length: 0
```

**Comment by pbohman** [comment]
short term fix #19265
